### PR TITLE
dl_iterate_phdr implementation

### PR DESCRIPTION
**Pull request by LemonBoy**
Also comes with a libc-less version that _should_ work just fine in most cases.
I guess the code is portable enough to work on *BSDs too, I'd be happy to move it out of the linux namespace if somebody tests it.

I didn't manage to mark the `__ehdr_entry` as weak, similar to how you'd do with this C declaration:
```c
extern const char __ehdr_start __attribute__((weak));
```

I'll squash the spurious commit once the review is over.

### CODE REVIEW

**Comment by andrewrk** [comment]
This looks good. I'm taking a look into the extern weak thing and see if I can resolve that, then I'll merge.

**Comment by andrewrk** [comment]
Ah - I do think it would be valuable to get some test coverage for this. Related:
https://github.com/ziglang/zig/blob/75d42ace31e8965a6bce04965dfc4a9a7b322b4f/test/build_examples.zig#L20-L26

File: std/c/linux.zig [hunk 1]
```diff
@@ -11,3 +11,5 @@ pub const pthread_attr_t = extern struct {
 
 /// See std.elf for constants for this
 pub extern fn getauxval(__type: c_ulong) c_ulong;
+
+pub extern fn dl_iterate_phdr(callback: *const c_void, data: ?*c_void) i32;
```
Comment (andrewrk) [inline std/c/linux.zig]: why not specify the type of the callback function here?

File: std/dynamic_library.zig [hunk 2]
```diff
@@ -19,6 +19,85 @@ pub const DynLib = switch (builtin.os) {
     else => void,
 };
 
+const LinkMap = extern struct {
+    l_addr: usize,
+    l_name: [*]u8,
+    l_ld: [*c]elf.Dyn,
```
Comment (andrewrk) [inline std/dynamic_library.zig]: All the C pointers in this file should be normal pointers.

File: std/dynamic_library.zig [hunk 3]
```diff
@@ -19,6 +19,85 @@ pub const DynLib = switch (builtin.os) {
...
+    l_next: [*c]LinkMap,
+    l_prev: [*c]LinkMap,
+
+    pub const Iterator = struct {
+        lm_ptr: [*c]LinkMap,
+
+        fn end(self: *const Iterator) bool {
```
Comment (andrewrk) [inline std/dynamic_library.zig]: This can probably be simply: 

```suggestion
        fn end(self: Iterator) bool {
```

File: std/os/linux.zig [hunk 4]
```diff
@@ -1534,6 +1535,68 @@ pub const dirent64 = extern struct {
     d_name: u8, // field address is the address of first byte of name https://github.com/ziglang/zig/issues/173
 };
 
+pub const dl_phdr_info = extern struct {
+    dlpi_addr: usize,
+    dlpi_name: [*c]const u8,
```
Comment (andrewrk) [inline std/os/linux.zig]: No C pointers, please!

File: std/os/linux.zig [hunk 5]
```diff
@@ -1534,6 +1535,68 @@ pub const dirent64 = extern struct {
...
+    dlpi_phdr: [*c]*c_void,
+    dlpi_phnum: u16,
+};
+
+const DynLib = @import("../dynamic_library.zig");
+
+// XXX: This should be weak
```
Comment (andrewrk) [inline std/os/linux.zig]: Is this PR ok to merge even before we solve the weak thing? What are the implications of this?

**Comment by andrewrk** [comment]
Some fix ups for you to look at, and then I do think we should have tests for this, and then I think it will be good to merge.

File: std/os/linux/test.zig [hunk 6]
```diff
@@ -42,3 +44,39 @@ test "timer" {
     // TODO implicit cast from *[N]T to [*]T
     err = linux.epoll_wait(@intCast(i32, epoll_fd), @ptrCast([*]linux.epoll_event, &events), 8, -1);
 }
+
+export fn iter_fn(info: *linux.dl_phdr_info, size: usize, data: ?*usize) i32 {
+    var counter = data orelse @panic("???");
```
Comment (daurnimator) [inline std/os/linux/test.zig]: Could just be `orelse unreachable` I think?

**Comment by LemonBoy** [comment]
> Is this PR ok to merge even before we solve the weak thing? What are the implications of this?

It should not matter since lld produces that symbol (and so does binutils `as`).

I've added a simple test for the non-libc case that should be quite exhaustive.

A few notes:
- Is the exposed API ergonomic enough? I've made it accept a `comptime T: type` parameter in order to make it more type-safe but that's it.
- We require the callback to be `extern` because we may call the libc function (IMO that's not needed, this implementation is 100% compatible with either libc and, if I didn't overlook anything, also with BSDs). This means the callback is effectively exported even though that's not needed, all we care for is the calling convention. Until #661 (and unless `@export` is extended to also allow the user to select the symbol visibility) there's no way around this.

File: std/os/linux/test.zig (hunk 6 above)
Comment (LemonBoy) [reply std/os/linux/test.zig]: Good catch, I went with `.?`

File: std/os/linux/test.zig [hunk 7]
```diff
@@ -42,3 +44,39 @@ test "timer" {
...
+    }
+
+    return 0;
+}
+
+test "dl_iterate_phdr" {
+    var counter: usize = 0;
+    const r = linux.dl_iterate_phdr(usize, iter_fn, &counter);
+    expect(r == 0);
+    expect(counter == 1);
```
Comment (daurnimator) [inline std/os/linux/test.zig]: Is this liable to break in future if we somehow end up using other libraries in the test suite?

Comment (LemonBoy) [reply std/os/linux/test.zig]: Yep, but the test is easy to change when/if that happens.

**Comment by andrewrk** [comment]
Test failure on the CI:

```
848/1106 os.linux.test.test "std-linux-x86_64-Debug-c-multi dl_iterate_phdr"...test failure
/home/vsts/work/1/s/std/testing.zig:146:14: 0x2ad418 in testing.expect (test)
/home/vsts/work/1/s/std/os/linux/test.zig:80:11: 0x3085a2 in os.linux.test.test "std-linux-x86_64-Debug-c-multi dl_iterate_phdr" (test)
/home/vsts/work/1/s/std/special/test_runner.zig:13:25: 0x6e8cca in std.special.main (test)
/home/vsts/work/1/s/std/special/bootstrap.zig:126:22: 0x6e8996 in std.special.main (test)
```

**Comment by LemonBoy** [comment]
> Test failure on the CI:

@daurnimator was right, the test was too flakey. I've replaced it with one inspired by [this one](https://android.googlesource.com/platform/bionic/+/master/tests/link_test.cpp) and now the CI is green :)

**Pull request merged**